lib/evaluate/integer.h

```diff
@@ -0,0 +1,853 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_EVALUATE_INTEGER_H_
+#define FORTRAN_EVALUATE_INTEGER_H_
+
+// Emulates binary integers of an arbitrary (but fixed) bit size for use
+// when the host C++ environment does not support that size or when the
+// full suite of Fortran's integer intrinsic scalar functions are needed.
+// The data model is typeless, so signed* and unsigned operations
+// are distinguished from each other with distinct member function interfaces.
+// (*"Signed" here means two's-complement, just to be clear.  Ones'-complement
+// and signed-magnitude encodings appear to be extinct in 2018.)
+
+#include "bit-population-count.h"
+#include "common.h"
+#include "leading-zero-bit-count.h"
+#include <cinttypes>
+#include <climits>
+#include <cstddef>
+#include <type_traits>
+
+namespace Fortran::evaluate::value {
+
+// Implements an integer as an assembly of smaller host integer parts
+// that constitute the digits of a large-radix fixed-point number.
+// For best performance, the type of these parts should be half of the
+// size of the largest efficient integer supported by the host processor.
+// These parts are stored in either little- or big-endian order, which can
+// match that of the host's endianness or not; but if the ordering matches
+// that of the host, raw host data can be overlaid with a properly configured
+// instance of this class and used in situ.
+// To facilitate exhaustive testing of what would otherwise be more rare
+// edge cases, this class template may be configured to use other part
+// types &/or partial fields in the parts.  The radix (i.e., the number
+// of possible values in a part), however, must be a power of two; this
+// template class is not generalized to enable, say, decimal arithmetic.
+// Member functions that correspond to Fortran intrinsic functions are
+// named accordingly so that they can be referenced easily in the
+// language standard.
+template<int BITS, bool IS_LITTLE_ENDIAN = IsHostLittleEndian,
+    int PARTBITS =
+        BITS<32 ? BITS : 32, typename PART = HostUnsignedInt<PARTBITS>,
+            typename BIGPART = HostUnsignedInt<PARTBITS * 2>> class Integer {
```

Comment:
clang-format is clearly baffled by this template. I suggest replacing `BITS<32` with `BITS<=32` and it will do a better job.